### DISCUSSION TITLE
Why do we use convertToFloat() function

### DISCUSSION THREAD

**Discussion by Zen4All** [zencart/zencart#5797] [Q&A]
In the updat_product.php file we use the function convertToFloat() to loke the function says convert the input to a float.
Why do we not use a simple typecast to a float? like
 ```php
(float)$variable;
```

**Comment by lat9** [comment]
A function like that would be especially helpful on the storefront, enabling the Zen Cart 'base' to provide some consistent currency-related conversion.

While I don't agree with the `convertToFloat` implementation (since that function ultimately returns a casted float if the values appear numeric), I agree with the principle.  

Perhaps a _common_ `zen_float` function should be introduced that includes the current implementation with a notification to enable an override (using sprintf, for instance) to use other character-to-currency conversion methods that don't result in occasional penny-off calculations due to the rounding that's inherent in float calculations.

**Comment by scottcwilson** [comment]
I think I'm tilting towards just using typecasting.  Using `convertToFloat` on a garbage value like "Zen1.5.7" returns 1.5 - does that even make sense?

**Comment by drbyte** [comment]
I added that in https://github.com/zencart/zencart/pull/1516

It was born from the desire to extract repeated functionality like:
```php
$tmp_value = zen_db_prepare_input($_POST['products_quantity']);
$products_price = (!zen_not_null($tmp_value) || $tmp_value=='' || $tmp_value == 0) ? 0 : $tmp_value;
```
... and others.

In hindsight given that the Sanitizer handles invalid characters, I guess the following is better than `convertToFloat()`, since it handles `isset()`, `null`, `''`, `0`, and casting:
```php
$products_price = empty($_POST['products_quantity']) ? 0 : (float)$_POST['products_quantity'];
```

> Perhaps a common zen_float function should be introduced that includes the current implementation with a notification to enable an override

Both overriding PHP functions and inserting Notifier hooks to do even more overrides feels like an antipattern, which I prefer to avoid.

**Comment by lat9** [comment]
> Both overriding PHP functions and inserting Notifier hooks to do even more overrides feels like an antipattern, which I prefer to avoid.

In this case, though, a notifier hook would enable alternate value-methods (like the sprintf I mentioned above) that could reduce the penny-off rounding issues.  That way, an alternate solution could be created (without core-file modifications).

**Comment by drbyte** [comment]
Your point is heard fully. 
It's still an anti-pattern which should be avoided as much as possible.

A penny-off solution should be part of core.

I feel like we're now straying far from the original topic, which is focused on handling input data when editing products.

**Comment by neekfenwick** [comment]
`convertToFloat` is only ever used in `update_product.php` where it is taking user entered data, which can be pretty unreliable.  Is the intention to strip extra words in the input and retain the numeric part? e.g.

```
convertToFloat("10 cans") => 10 (double)
(float)"10 cans" => 0 (double)
```

So it's useful if the user copy/pastes some rubbish into the Create/Update Product form, the system keeps the numeric part and ignores any fluff.
